File: Data_extraction/extraction_scripts/java8/src/main/java/org/tonde/Clients.java
package org.tonde;

import com.datastax.oss.driver.api.core.CqlSession;
import com.datastax.oss.driver.api.core.cql.SimpleStatement;


import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Clients {
    private final CqlSession session;
    // Le nom de l'espace de clés
    private final String $KEYSPACE = "bigdata_exam";  // Changez ceci avec votre espace de clés
    // Le fichier contenant les données des clients
    private final String $FILE = "Clients.csv";  // Changez ceci avec le chemin de votre fichier CSV

    public static void main(String args[]) {
        try {
            Clients arb = new Clients();
            arb.initClientsTablesAndData(arb);
        } catch (RuntimeException e) {
            System.out.println("Erreur : " + e);
        }
    }

    /**
     * Constructeur
     */
    Clients() {
        session = CqlSession.builder()
                .withKeyspace($KEYSPACE)
                .build();
    }

    /**
     * Initialise la table des clients et charge les données à partir d'un fichier
     *
     * @param arb l'objet Clients
     */
    public void initClientsTablesAndData(Clients arb) {
        arb.dropTableClients();
        arb.createTableClients();
        arb.loadClientsDataFromFile($FILE);
    }

    /**
     * Supprime la table Clients
     */
    public void dropTableClients() {
        String statement = "DROP TABLE IF EXISTS Clients";
        executeDDL(statement);
    }

    /**
     * Crée la table Clients
     */
    public void createTableClients() {
        String statement = "CREATE TABLE Clients ("
                + "CLIENTID int PRIMARY KEY,"
                + "AGE int,"
                + "SEXE text,"
                + "TAUX int,"
                + "SITUATION_FAMILIALE text,"
                + "NBR_ENFANT int,"
                + "VOITURE_2 text,"
                + "IMMATRICULATION text)";
        executeDDL(statement);
    }

    /**
     * Charge les données des clients à partir d'un fichier
     *
     * @param clientDataFileName le nom du fichier contenant les données des clients
     */
    void loadClientsDataFromFile(String clientDataFileName) {
        InputStreamReader ipsr;
        BufferedReader br = null;
        InputStream ips;
        String ligne;

        System.out.println("Chargement des clients...");

        try {
            ips = Files.newInputStream(Paths.get(clientDataFileName));
            ipsr = new InputStreamReader(ips);
            br = new BufferedReader(ipsr);
            int i = 0;
            while ((ligne = br.readLine()) != null) {
                ArrayList<String> clientRecord = new ArrayList<>();
                StringTokenizer val = new StringTokenizer(ligne, ",");
                i += 1;
                while (val.hasMoreTokens()) {
                    clientRecord.add(val.nextToken());
                }
                if (clientRecord.get(1).equals("N/D")
                        || clientRecord.get(3).equals("N/D")
                        || clientRecord.get(5).equals("N/D") || clientRecord.get(0).equals("ClientId") || i == 1) {
                    continue;
                }

                int clientid = Integer.parseInt(clientRecord.get(0)) - 1;
                int age = Integer.parseInt(clientRecord.get(1));
                String sexe = clientRecord.get(2);
                int taux = Integer.parseInt(clientRecord.get(3));
                String SFamiliale = clientRecord.get(4);
                int nbEnfantsAcharge = Integer.parseInt(clientRecord.get(5));
                String voiture_2 = clientRecord.get(6);
                String immatriculation = clientRecord.get(7);
                this.insertAClientRow(clientid, age, sexe, taux, SFamiliale, nbEnfantsAcharge, voiture_2, immatriculation);
            }
        } catch (IOException e) {
            System.out.println("Erreur : " + e);
        }
    }

    /**
     * Insère une ligne dans la table Clients
     *
     * @param clientid          l'identifiant du client
     * @param age               l'age du client
     * @param sexe              le sexe du client
     * @param taux              le taux du client
     * @param SFamiliale        la situation familiale du client
     * @param nbEnfantsAcharge  le nombre d'enfants à charge du client
     * @param voiture_2         la voiture du client
     * @param immatriculation   l'immatriculation du client
     */
    private void insertAClientRow(int clientid, int age, String sexe, int taux, String SFamiliale, int nbEnfantsAcharge,
                                  String voiture_2, String immatriculation) {
        String statement = "INSERT INTO Clients (CLIENTID, AGE, SEXE, TAUX, SITUATION_FAMILIALE, NBR_ENFANT, VOITURE_2, IMMATRICULATION) " +
                "VALUES (?, ?, ?, ?, ?, ?, ?, ?)";
        session.execute(SimpleStatement.builder(statement)
                .addPositionalValue(clientid)
                .addPositionalValue(age)
                .addPositionalValue(sexe)
                .addPositionalValue(taux)
                .addPositionalValue(SFamiliale)
                .addPositionalValue(nbEnfantsAcharge)
                .addPositionalValue(voiture_2)
                .addPositionalValue(immatriculation)
                .build());
    }

    /**
     * Exécute une instruction DDL
     *
     * @param statement l'instruction DDL à exécuter
     */
    public void executeDDL(String statement) {
        session.execute(statement);
    }
}

File: BigData/extraction_scripts/insertion_client_marketing/clients.py
from cassandra.cluster import Cluster
from cassandra.query import SimpleStatement
import csv


class Clients:
    def __init__(self, keyspace, file_path, container_name="cassandra-container"):
        self.keyspace = keyspace
        self.file_path = file_path
        self.cluster = Cluster([container_name])
        self.session = self.cluster.connect(self.keyspace)

    def init_clients_tables_and_data(self):
        self.drop_table_clients()
        self.create_table_clients()
        self.load_clients_data_from_file(self.file_path)

    def drop_table_clients(self):
        statement = "DROP TABLE IF EXISTS Clients"
        self.execute_ddl(statement)

    def create_table_clients(self):
        statement = """
            CREATE TABLE Clients (
                CLIENTID int PRIMARY KEY,
                AGE int,
                SEXE text,
                TAUX int,
                SITUATION_FAMILIALE text,
                NBR_ENFANT int,
                VOITURE_2 text,
                IMMATRICULATION text
            )
        """
        self.execute_ddl(statement)

    def load_clients_data_from_file(self, file_path):
        print("Chargement des clients...")
        try:
            with open(file_path, mode='r') as file:
                print('Reading file...')
                reader = csv.reader(file)
                print('Skipping header row...')
                i = 0
                for row in reader:
                    i += 1
                    try:
                        print('ROW: ', row)
                        if row[1] == "N/D" or row[3] == "N/D" or row[5] == "N/D" or i == 1:
                            continue
                        clientid = int(row[0])
                        print('ID is clean')
                        age = int(row[1])
                        print('AGE is clean')
                        sexe = row[2]
                        print('SEXE is clean')
                        taux = int(row[3])
                        print('TAUX is clean')
                        s_familiale = row[4]
                        print('SITUATION_FAMILIALE is clean')
                        nb_enfants_a_charge = int(row[5])
                        print('NBR_ENFANT is clean')
                        voiture_2 = row[6]
                        print('VOITURE_2 is clean')
                        immatriculation = row[7]
                        print('IMMATRICULATION is clean')
                        self.insert_a_client_row(
                            clientid, age, sexe, taux, s_familiale, nb_enfants_a_charge, voiture_2, immatriculation)
                        print('Row inserted')
                    except ValueError as e:
                        print(f"Error converting row: {row}. Error: {e}")
        except IOError as e:
            print("Erreur : ", e)

    def insert_a_client_row(self, clientid, age, sexe, taux, s_familiale, nb_enfants_a_charge, voiture_2, immatriculation):
        statement = """
            INSERT INTO Clients (CLIENTID, AGE, SEXE, TAUX, SITUATION_FAMILIALE, NBR_ENFANT, VOITURE_2, IMMATRICULATION) 
            VALUES (%s, %s, '%s', %s, '%s', %s, '%s', '%s')
        """ % (clientid, age, sexe, taux, s_familiale, nb_enfants_a_charge, voiture_2, immatriculation)

        print('Creating SimpleStatement...')
        print('Statement: ', statement)
        # Créer un objet SimpleStatement à partir de la chaîne de requête
        simple_statement = SimpleStatement(statement)
        print('SimpleStatement created')
        # Passer uniquement l'objet SimpleStatement à la méthode execute
        self.session.execute(simple_statement)


    def execute_ddl(self, statement):
        self.session.execute(statement)


if __name__ == "__main__":
    try:
        clients = Clients(keyspace="bigdata_exam", file_path="Client.csv")
        clients.init_clients_tables_and_data()
    except Exception as e:
        print("Erreur : ", e)
